### ISSUE TITLE
Add Batch Normalization to Basic_CNN_128x128

### ISSUE THREAD

**Issue by alexwdong** [alexwdong/IncubatorCVProject#19]
Basic_CNN_128x128 has four convolutional layers, each with 128 channels. This is a really complex model with a lot of parameters, and i can't get the performance to improve) (performance doesn't improve after each epoch, despite training for 100 epochs) . Can we try adding batch normalization/maybe dropout to see if we can get it to train?

**Comment by TRokieG** [comment]
I saw in the code that we have drop out already. Do we need to change our drop out code as well?

**Comment by TRokieG** [comment]
In addition, I don't know for sure, but when I take a look at our dog dataset data (print out the vectors), there are many zeros in most of the matrix. Is this normal? Do we need to check our data-loader as well ? 
![image](https://user-images.githubusercontent.com/36273798/86983623-5aa1ec00-c15a-11ea-97d0-251f7bba19ad.png)

**Comment by alexwdong** [comment]
The zeros might be coming from the fact that the images are padded to make them square.

I haven't tried training with dropout, but as far as I understand, dropout is used for regularising the neural net (making it perform better in the validation set). The issue I'm facing is that the neural net won't even learn in the first place, so I'm guessing that batch normalization will help with that

**Comment by TRokieG** [comment]
Got it. I am not very familiar with drop out and batch normalization, but I will read some blog posts and try them to see if there is any performance improvement. In addition, with more layers adding, there is no improvement at all ?

**Comment by TRokieG** [comment]
I am thinking that probably we should not use drop out, because to my understanding, drop out is used to prevent overfitting. Our main issue right now is very low accuracy, both for training set and validation set. If we add drop out in, probably we will have even lower accuracy. So my main focus now is to add more layers, add batch normalization and tune hyper parameters. So far I have tried adding batch normalization at the end of the last Conv layer, but it still didn't generate a higher accuracy....

**Comment by alexwdong** [comment]
Try adding batch normalization after each convolutional layer, but before the activation.

So one 'stack' of convolutional layers would be :
Convolution->batchnorm->activation->maxpool

**Comment by TRokieG** [comment]
I tried adding batch normalization into the network, but the result is still not ideal. Now I tried to add in more Conv layers and fully connected layers, but I always run into size mismatch problems. Alex is there a blog or some straightforward universal rules that I can follow to make sure size mismatch won't happen ? 
![image](https://user-images.githubusercontent.com/36273798/87177887-5c7ac500-c2aa-11ea-9577-d2492a41892d.png)

**Comment by alexwdong** [comment]
I haven't come across a good blog post or resource that details the convolution dimensions. What I've been doing is using a convolution with padding =1, stride=1, and kernel_size = 3. This convolution will take any _n_ x _n_ matrix and return another _n_ x _n_ matrix (of the exact same size). Then, after using a max pool with 2x2 kernel size, the dimension _n_ x _n_ matrix will be reduced to _n/2_ x _n/2_. Then I usually pick n so that its of form 2^k (e.g, 8,16,32,64,128)

**Comment by alexwdong** [comment]
I guess the best thing to do is just try to be careful with how convolutions affect how dimensions work. When I mean that I haven't come across a blog post, it means that I havent come across some formula that takes padding,stride,kernel size, and input dimension, and gives you an output dimension

**Comment by alexwdong** [comment]
Also, can you commit some of your code so I can take a look at it?

**Comment by TRokieG** [comment]
No problem. I am still trying different combinations of parameters/layers. Let me try a few more times and I'll upload the new ones up to Github.

**Comment by TRokieG** [comment]
I already uploaded the code I wrote on Combine branch. It is in src/Basic_CNN_Architecture.py, with name 'BasicCNN_128x128_w_bn'. I tired different layers and different parameters, but none of them worked well. Instead, the highest validation score I got is 5%, with the same model as the original one you have and add batch normalization after two Convnets .  I am thinking the reason why our model doesn't work well might have something to do with how we preprocess the data? I saw on kaggle that one uploaded kernel actually crops all the images first to only contain dogs, and then start training. They can achieve accuracy of over 70%. Or maybe 128*128 images are too blurry ? I don't know...

**Comment by alexwdong** [comment]
Ok, so, your first two convolutions have stride=2, i'm not sure if this is deliberate, but this will definitely reduce the dimensions of the feature map after each of those convolutions, which explains why your the dimensions of the feature map are going to 0 x 0. 

In addition, (i'm not 100% sure on this), but batch normalization won't increase the "best performance" of the neural network. Rather, it will make training your neural network easier.

In the NN model you committed, you only have 8 channels in each convolutional layer, which is probably not enough complexity for batch normalization to be needed. Try increasing the number of channels in each layer to 128.
